quantization/src/encoded_vectors_binary.rs

```diff
@@ -0,0 +1,188 @@
+        debug_assert!(validate_vector_parameters(orig_data.clone(), vector_parameters).is_ok());
+
+        for vector in orig_data {
+            if stop_condition() {
+                return Err(EncodingError::Stopped);
+            }
+
+            let encoded_vector = Self::_encode_vector(vector);
+            let encoded_vector_slice = encoded_vector.encoded_vector.as_slice();
+            let bytes = transmute_to_u8_slice(encoded_vector_slice);
+            storage_builder.push_vector_data(bytes);
+        }
+
+        Ok(Self {
+            encoded_vectors: storage_builder.build(),
+            metadata: Metadata {
+                vector_parameters: vector_parameters.clone(),
+            },
+        })
+    }
+
+    fn _encode_vector(vector: &[f32]) -> EncodedBinVector {
+        let mut encoded_vector = vec![0; Self::get_storage_size(vector.len())];
+
+        for (i, &v) in vector.iter().enumerate() {
+            // flag is true if the value is positive
+            // It's expected that the vector value is in range [-1; 1]
+            if v > 0.0 {
+                encoded_vector[i / BITS_STORE_TYPE_SIZE] |= 1 << (i % BITS_STORE_TYPE_SIZE);
+            }
+        }
+
+        EncodedBinVector { encoded_vector }
+    }
+
+    /// Xor vectors and return the number of bits set to 1
+    ///
+    /// Assume that `v1` and `v2` are aligned to `BITS_STORE_TYPE_SIZE` with both with zeros
+    /// So it does not affect the resulting number of bits set to 1
+    fn xor_product(v1: &[BitsStoreType], v2: &[BitsStoreType]) -> usize {
+        debug_assert!(v1.len() == v2.len());
+        let mut result = 0;
+        for i in 0..v1.len() {
+            result += (v1[i] ^ v2[i]).count_ones() as usize;
+        }
+        result
+    }
+
+    /// Estimates how many `StorageType` elements are needed to store `size` bits
+    fn get_storage_size(size: usize) -> usize {
+        let mut result = size / BITS_STORE_TYPE_SIZE;
+        if size % BITS_STORE_TYPE_SIZE != 0 {
+            result += 1;
+        }
+        result
+    }
+
+    /// Get the size of the whole vector storage in bytes
+    fn get_quantized_vector_size(vector_parameters: &VectorParameters) -> usize {
+        let actual_dim = Self::get_vector_size_bytes_from_params(vector_parameters);
+        actual_dim * vector_parameters.count
+    }
+
+    fn get_vector_size_bytes_from_params(vector_parameters: &VectorParameters) -> usize {
+        Self::get_storage_size(vector_parameters.dim) * std::mem::size_of::<BitsStoreType>()
+    }
+
+    fn get_vector_size_bytes(&self) -> usize {
+        Self::get_vector_size_bytes_from_params(&self.metadata.vector_parameters)
+    }
+
+    fn calculate_metric(&self, v1: &[BitsStoreType], v2: &[BitsStoreType]) -> f32 {
+        let xor_product = Self::xor_product(v1, v2);
+
+        // Dot product in a range [-1; 1] is approximated by NXOR in a range [0; 1]
+        // For example:
+
+        // A    |   B   | Dot product
```

Comment:
We can also use `[0; 1]` range with AND operation. I propose to support both operations

```
A    |   B   | Dot product
0 | 0 |  0
0 |  1 | 0
1 | 0  | 0
1 |  1  |  1
```

Reply:
let's estimate on real embeddings before actually implementing